File: lab1/Drink.cpp
#include "Drink.h"
#include <iostream>
using namespace std;

// конструктор
Drink::Drink(string name, int volume, int price){
	this->name = name; 
	this->volume = volume;
	this->price = price;
	// this->name – это поле
	// name – это параметр
	// иначе компилятор не видит разницы
}


// геттеры
string Drink::GetName(){
	return name;
}

int Drink::GetVolume(){
	return volume;
}

int Drink::GetPrice(){
	return price;
}


// сеттеры
void Drink::SetName(string n){
	name = n;
}

void Drink::SetVolume(int v){
	volume = (v > 0) ? v : (cout << "Ошибка, объём должен быть больше 0" << endl, volume);
}

void Drink::SetPrice(int p){
	price = (p > 0) ? p : (cout << "Ошибка, цена должна быть больше 0" << endl, price);
}


// методы
void Drink::Info(){
	cout << "Название напитка: " << name << ", объём: " 
		 << volume << "мл" << ", цена: " << price << endl;
}


// деструктор
Drink::~Drink(){
	cout << "\nУдаляем напиток " << GetName() << endl;
}

File: laba2/models/Drink.cs
namespace Laba2App.Models;

public abstract class Drink
{
    // поля класса

    private int _id;
    private string _name;
    private int _volume;
    private int _price;

    // свойства

    public int Id
    {
        get => _id;
        set => _id = value;
    }

    public string Name
    {
        get => _name;
        set
        {
            if (!string.IsNullOrWhiteSpace(value))
                _name = value;
        }
    }

    public int Volume
    {
        get => _volume;
        set
        {
            if (value > 0)
                _volume = value;
        }
    }

    public int Price
    {
        get => _price;
        set
        {
            if (value > 0)
                _price = value;
        }
    }

    // полиморфные свойства

    public abstract string Type { get; }

    public abstract string ExtraInfo { get; }

    public string InfoText => Info();

    // конструктор

    protected Drink(string name, int volume, int price)
    {
        Name = name;
        Volume = volume;
        Price = price;
    }

    // виртуальный метод информации

    public virtual string Info()
    {
        return $"name: {Name}, volume: {Volume} ml, price: {Price}";
    }

    // абстрактный метод

    public abstract string GetWishes();

    // деструктор

    ~Drink()
    {
        System.Diagnostics.Debug.WriteLine($"deleting drink {Name}");
    }
}